### PR TITLE
Update diagnostic to make it more general

### PR DESCRIPTION
**Pull request by MSECode**
Make for abs encoders (at joint) are more general
Moroever, now we show the abs encoder type too in the diagnostic message, solving the issue: https://github.com/robotology/icub-firmware/issues/533
Related to: 
- https://github.com/robotology/icub-firmware-shared/pull/102
- https://github.com/robotology/icub-firmware/pull/540

### CODE REVIEW

**Comment by pattacini** [comment]
Hi @MSECode 

Remember to manage versioning.

File: CMakeLists.txt
Comment (pattacini) [inline CMakeLists.txt:9]: Unless strictly required please leave it unchanged

**Review by pattacini — CHANGES_REQUESTED** [review]

File: CMakeLists.txt
Comment (MSECode) [reply CMakeLists.txt:9]: Oks, I'll leave the version to 2.7.1 then. So, which versioning should I care about. Just icub-firmware-shared?

Comment (pattacini) [reply CMakeLists.txt:9]: > So, which versioning should I care about. Just icub-firmware-shared and update conf/iCubFindDependencies.cmake accordingly and that's all?

Yep, thanks!

**Review by marcoaccame — CHANGES_REQUESTED** [review]

File: src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp
Comment (MSECode) [reply src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp:170]: I'll update the code as suggested then, thanks

Comment (MSECode) [reply src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp]: Hi @marcoaccame, 
I've applied the suggested modifications, also updating the CMakeFIleLists.txt to have a cleaner include of embObjMotionControl. 
Just one one, I think that the `reinterpret_cast` is having some unexpected behavior. As a matter of fact, when the method in embObjMC is called the core dumps due to a segmentation fault (thus we have issues in the memory allocation) as here:
```
[INFO] All actions for action level 20 of startup phase finished.
[INFO] startup phase finished.
[INFO] run phase starting...
[DEBUG] yarprobotinterface running happily
[ERROR] yarprobotinterface intercepted a segmentation fault caused by a faulty plugin:
[ERROR] /home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN4yarp3dev19embObjMotionControl18getEncoderTypeNameEj15eOmc_position_tRNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0xa2) [0x7fd3ec06a352]
Trace requested at /home/iiticublap258/Workspace/robotology-superbuild/src/YARP/src/yarprobotinterface/Module.cpp:77 by code called from:
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libYARP_os.so.3(_Z16yarp_print_traceP8_IO_FILEPKcj+0x3b) [0x7fd3f5eb208b]
yarprobotinterface(+0x9cc6) [0x5644f0831cc6]
/lib/x86_64-linux-gnu/libc.so.6(+0x43090) [0x7fd3f5a56090]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN4yarp3dev19embObjMotionControl18getEncoderTypeNameEj15eOmc_position_tRNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE+0xa2) [0x7fd3ec06a352]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN10Diagnostic8LowLevel19MotionControlParser9parseInfoEv+0x624) [0x7fd3ec0eae74]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN10Diagnostic8LowLevel13InfoFormatter17getDiagnosticInfoERNS_12EmbeddedInfoE+0x4e8) [0x7fd3ec0e7f08]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(feat_manage_diagnostic+0x10a) [0x7fd3ec0e1cca]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libembobj.so.1.41.1(eo_nv_hid_remoteSetROP+0xc6) [0x7fd3e47da696]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libembobj.so.1.41.1(eo_agent_InpROPprocess+0x334) [0x7fd3e47d9324]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libembobj.so.1.41.1(eo_receiver_Process+0x1a4) [0x7fd3e47dc204]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libembobj.so.1.41.1(eo_transceiver_Receive+0x85) [0x7fd3e47ddc95]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN3eth15HostTransceiver8parseUDPEPKvt+0xa0) [0x7fd3ec0e3770]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN3eth13TheEthManager9ReceptionEjPml+0xad) [0x7fd3ec0ab59d]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/iCub/embObjMotionControl.so(_ZN3eth11EthReceiver3runEv+0x10b) [0x7fd3ec0b76fb]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libYARP_os.so.3(+0xb53be) [0x7fd3f5ed93be]
/home/iiticublap258/Workspace/robotology-superbuild/build/install/lib/libYARP_os.so.3(+0x17c08d) [0x7fd3f5fa008d]
/lib/x86_64-linux-gnu/libstdc++.so.6(+0xd6df4) [0x7fd3f5cf8df4]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x8609) [0x7fd3f5767609]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x43) [0x7fd3f5b32353]
Segmentation fault (core dumped)
```

Looking on the forums, I saw that it is suggested the safer use of `dynamic_cast`.
Substituting it in the casting here:
` if(embObjMotionControl *emc = dynamic_cast<embObjMotionControl*>(m_MC_ethRes); emc != nullptr)`

I got the expected result in the logs:

```
[DEBUG] yarprobotinterface running happily
[ERROR] from BOARD 10.0.1.1 (three-encoders-setup) time=3366s 36m 60u :  MC: Absolute encoder invalid data. Hardware problem in the magnetic position sensor of the joint caused invalid position readings. (Joint=three_encoders_joint (NIB=0), encoderPort=0, encoderType=AMO)
[ERROR] from BOARD 10.0.1.1 (three-encoders-setup) time=3371s 36m 60u :  MC: Absolute encoder invalid data. Hardware problem in the magnetic position sensor of the joint caused invalid position readings. (Joint=three_encoders_joint (NIB=0), encoderPort=0, encoderType=AMO)
[WARNING] from BOARD 10.0.1.1 (three-encoders-setup) time=3374s 509m 992u :  MC: Absolute encoder has spikes. There is impulsive noise in the measures of the magnetic position sensor of the joint. (Joint=three_encoders_joint (NIB=0), encoderPort=0, encoderType=AMO)
[ERROR] from BOARD 10.0.1.1 (three-encoders-setup) time=3376s 37m 65u :  MC: Absolute encoder invalid data. Hardware problem in the magnetic position sensor of the joint caused invalid position readings. (Joint=three_encoders_joint (NIB=0), encoderPort=0, encoderType=AMO)

```

Since, I'm not sure if a dynamic memory allocation is desirable in our code structure, what do u think about this change, thanks

Comment (marcoaccame) [reply src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp]: Hi @MSECode, my bad: `dynamic_cast<>` surely is the one to use. See https://en.cppreference.com/w/cpp/language/dynamic_cast where it says `Safely converts pointers and references to classes up, down, and sideways along the inheritance hierarchy.`

Then:

> Since, I'm not sure if a dynamic memory allocation is desirable in our code structure, what do u think about this change, thanks

it does not allocate new memory, it just correctly uses a pointer to an existing class.

Comment (marcoaccame) [reply src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp]: In conclusion:
very good of yours to have tested the changes and made them work.
As you have achieved the correct result in the log everything is fine by me

**Review by marcoaccame — APPROVED** [review]
As you have achieved the correct result in the log everything is fine by me

File: src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp
Comment (traversaro) [reply src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp]: I am not sure if this method is frequently called, but we had a few problems with performances in the past with `dynamic_cast` (see https://github.com/robotology/icub-main/pull/82). If the `m_MC_ethRes` pointer never changes, perhaps we can think of calling `dynamic_cast` only once in the constructor and storing its value in a class attribute?

Comment (MSECode) [reply src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp]: Since it is called quite often (depending on the number of error on the abs encoder), I'll do as you suggest. It makes sense to call the cast just one time in the constructor.

File: src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp
Comment (Nicogene) [inline src/libraries/icubmod/embObjLib/diagnosticInfoFormatter.cpp:171]: To avoid circular dependency I would:
- add `getEncoderTypeName` in the [`IEthResource`](https://github.com/robotology/icub-main/blob/22c06ceffe66c21de446da841a1ffd432ad515d0/src/libraries/icubmod/embObjLib/IethResource.h) with a default implementation
- Implement (marking it `override`) `getEncorderTypeName` in `embObjMotionControl`
- Change this code in order to directly invoke `m_MC_ethRes->getEncoderTypeName` without the need of the dynamic_cast

**Review by Nicogene — CHANGES_REQUESTED** [review]
Some suggestions for resolving the CI failure

cc @marcoaccame

**Review by pattacini — APPROVED** [review]

**Comment by pattacini** [comment]
The CI is now happy.
cc @Nicogene @valegagge @marcoaccame

**Review by Nicogene — APPROVED** [review]

**Comment by pattacini** [comment]
Awaiting the final approval of @marcoaccame and @valegagge before merging.

**Comment by pattacini** [comment]
/remind December 16, please let's proceed asap with this.

**Review by valegagge — APPROVED** [review]
LGTM

**Comment by pattacini** [comment]
Great! Let me merge it then as we need to address this failure:
- https://github.com/robotology/robotology-superbuild/issues/1764

Also, #989 will have to be rebased on top of this because of versioning.

**Pull request merged**
